Optimize ST_Envelope

mbasmanova opened the pull request:
Similar to #9798, ST_Envelope spent most of the time deserializing geometry from Slice. This PR updates ST_Envelope to limit deserialization to just the envelope.

Benchmark results before the change:

```
Benchmark                                      Mode  Cnt       Score       Error  Units
BenchmarkSTEnvelope.complexGeometry  avgt   20  173054.303 ± 24902.172  ns/op
BenchmarkSTEnvelope.simpleGeometry   avgt   20    1866.689 ±   388.987  ns/op
```

and after:

```
Benchmark                                      Mode  Cnt     Score    Error  Units
BenchmarkSTEnvelope.complexGeometry  avgt   20   977.294 ± 55.636  ns/op
BenchmarkSTEnvelope.simpleGeometry   avgt   20  1016.174 ± 57.284  ns/op
```

Also, fixes #9814

nezihyigitbasi commented on presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java:
```diff
@@ -392,10 +393,11 @@ public static Slice stBoundary(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
+        if (envelope == null) {
+            return EMPTY_POLYGON;
+        }
+        return serialize(createFromEsriGeometry(envelope, null));
```
Previously, you were passing the spatial reference of `geometry` to `createFromEsriGeometry()`, but now you pass `null`. Why is that change needed?

mbasmanova replied:
I no longer have the spatial reference of `geometry`, hence, cannot pass it. However, we don't use spatial reference anywhere and all geometries have `null` spatial reference because of `geometry.setSpatialReference(null);` in `geometryFromText`.

nezihyigitbasi commented on presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/BenchmarkSTEnvelope.java:
```diff
@@ -0,0 +1,83 @@
+
+@State(Scope.Thread)
+@Fork(2)
+@Warmup(iterations = 10, time = 500, timeUnit = TimeUnit.MILLISECONDS)
```
static import `TimeUnit.MILLISECONDS`

nezihyigitbasi commented on presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/BenchmarkSTEnvelope.java:
```diff
@@ -0,0 +1,83 @@
+@Fork(2)
+@Warmup(iterations = 10, time = 500, timeUnit = TimeUnit.MILLISECONDS)
+@Measurement(iterations = 10, time = 500, timeUnit = TimeUnit.MILLISECONDS)
+@OutputTimeUnit(TimeUnit.NANOSECONDS)
```
static import `TimeUnit.NANOSECONDS`

nezihyigitbasi commented on presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/GeometryBenchmarkUtils.java:
```diff
@@ -0,0 +1,35 @@
+/*
```
Please add this refactoring as a separate commit.

nezihyigitbasi commented on presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/BenchmarkSTEnvelope.java:
```diff
@@ -0,0 +1,83 @@
+public class BenchmarkSTEnvelope
+{
+    @Benchmark
+    public Object stEnvelopeSimpleGeometry(BenchmarkData data)
```
The name of this class already contains `STEnvelope` so you can simplify the names of the benchmark methods by removing that (`simpleGeometry`/`complexGeometry`).

nezihyigitbasi commented on presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/BenchmarkSTEnvelope.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    @Benchmark
+    public Object stEnvelopeComplexGeometry(BenchmarkData data)
```
You can return `Slice` from these two benchmark methods.

mbasmanova commented:
@nezihyigitbasi Nezih, thank you for review. I updated the PR to incorporate your comments.

nezihyigitbasi commented on presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/GeometryBenchmarkUtils.java:
```diff
@@ -0,0 +1,35 @@
+{
+    private GeometryBenchmarkUtils() {}
+
+    public static String loadLargePolygon()
```
nit: We can update the signature of the method as `loadPolygon(String fileName)` instead of tying it to a certain file.

nezihyigitbasi commented:
The failing test seems unrelated, and it was previously reported as flaky.

mbasmanova commented:
@nezihyigitbasi Thank you, Nezih. I addressed the comments and updated the PR. For some reason I don't see the Travis job kicking in. Do you happen to know why? I'd like to wait for the test results before merging.

nezihyigitbasi commented:
There is a release going on right now, please talk to @raghavsethi before merging if this needs to be merged.
